### DISCUSSION TITLE
Lamps with relay

### DISCUSSION THREAD

**Discussion by chrmuh** [HomeSpan/HomeSpan#930] [Q&A]
Hi,

I am looking forward to start a new project with homespan. But I do not realy know which Code I should use.
I want to control a lamp of my house with a relay that is controlled by a button or a pin of an esp. On the other hand there should also be a pin that monitors the state of the relay and depending on that state it shold show the state of the lamp in homekit

Which code parts do I need to achive that idea?

Best wishes

**Comment by HomeSpan** [comment]
See the PushButton Tutorial Example for a good start.

**Comment by delingren** [comment]
I have some code that does exactly that. It's rather straightforward. You need an output pin for controlling the relay and an input pin for the pushbutton. Here's my code. The relay pin triggers on high and the pushbutton pin triggers on low (shorted to GND):

```
// A switch that controls a relay but can also be manually operated by a
// pushbutton. It uses two pins. One output pin for controlling the relay and
// one input pin for the pushbutton. If the pushbutton pin doesn't have an
// internal pullup resistor, use an external one.
struct DEV_Switch : Service::Switch {
  int pin_relay;
  int pin_button;
  SpanCharacteristic *power;

  DEV_Switch(int pin_relay, int pin_button) : Service::Switch() {
    this->pin_relay = pin_relay;
    this->pin_button = pin_button;
    power = new Characteristic::On();
    pinMode(pin_relay, OUTPUT);

    new SpanButton(pin_button, PushButton::TRIGGER_ON_LOW);
  }

  boolean update() {
    digitalWrite(pin_relay, power->getNewVal());
    return true;
  }

  void button(int pin, int type) override {
    if (pin != pin_button) {
      return;
    }
    bool new_value = !(power->getNewVal());
    power->setVal(new_value);
    digitalWrite(pin_relay, new_value);
  }
};
```
